libcxx/include/__type_traits/is_trivially_relocatable.h

```diff
@@ -0,0 +1,51 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___TYPE_TRAITS_IS_TRIVIALLY_RELOCATABLE_H
+#define _LIBCPP___TYPE_TRAITS_IS_TRIVIALLY_RELOCATABLE_H
+
+#include <__config>
+#include <__type_traits/is_trivially_destructible.h>
+#include <__type_traits/is_trivially_move_constructible.h>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+//------------------------------------------------------WARNING-------------------------------------------------------//
+//
+// This type trait should only be used where ABI stability isn't relevant or can be assured by other means,
+// like in function bodies or where the difference can be detected at runtime. If this trait is used in ABI-sensitive
+// environments it has to be ensured that Clang and GCC can interpret the data produced by each other. If you are not
+// certain that this is the case DO NOT USE THIS TYPE TRAIT
+//
+//--------------------------------------------------------------------------------------------------------------------//
```

Comment:
I don't think `__libcpp_is_trivially_relocatable` is a good idea if it needs big scary code-comments about ABI-stability. What happens when Clang 16 changes the meaning of the builtin so that e.g. `__is_trivially_relocatable(reference_wrapper<int()>)` used to be false but now it's true? Does that break libc++'s `move_only_function` ABI? I actually think the answer is "no, it doesn't break ABI; linking old and new TUs together will certainly violate the ODR, but there's no specific ABI concern here." But in that case, I'm not sure why the big scary comment is needed.

Reply:
> I don't think `__libcpp_is_trivially_relocatable` is a good idea if it needs big scary code-comments about ABI-stability.

The big scary comment is to avoid accidentally depending on it in an ABI context, since it's not stable.

> What happens when Clang 16 changes the meaning of the builtin so that e.g. `__is_trivially_relocatable(reference_wrapper<int()>)` used to be false but now it's true? 

Essentially nothing. The destruction function does something different.

> Does that break libc++'s `move_only_function` ABI? I actually think the answer is "no, it doesn't break ABI; linking old and new TUs together will certainly violate the ODR, but there's no specific ABI concern here."

Assuming you use releases, the answer is actually "No, it neither breaks ABI nor violates the ODR". Every `HIDE_FROM_ABI` function now has an ABI tag, so there aren't any ODR violations linking old and new TUs. See https://reviews.llvm.org/D127444 for more context.

> But in that case, I'm not sure why the big scary comment is needed.

It's not meant for `move_only_function`. It's meant for cases where it would be ABI breaking. I just want to make sure that people using `is_trivially_relocatable` realize that it's not ABI stable.